File: osmToolsBasic/src/main/java/com/asamm/osmTools/mapConfig/ItemMap.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package com.asamm.osmTools.mapConfig;

import com.asamm.osmTools.config.Action;
import com.asamm.osmTools.sea.Boundaries;
import com.asamm.osmTools.utils.Logger;
import com.asamm.osmTools.utils.Utils;
import lombok.Getter;
import lombok.Setter;
import net.minidev.json.JSONObject;
import net.minidev.json.parser.JSONParser;
import net.minidev.json.parser.ParseException;
import org.apache.commons.io.FileUtils;
import org.kxml2.io.KXmlParser;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.nio.file.Path;

import static java.nio.charset.StandardCharsets.UTF_8;

/**
 * @author volda
 */
public class ItemMap extends AItemMap {

    private static final String TAG = ItemMap.class.getSimpleName();

    // BASIC PARAMETERS

    // unique ID of item
    @Getter
    private String id;
    // name of file
    @Getter
    private String name;
    // how will item called in the store
    @Getter
    private String nameReadable;

    // name of item for generating (useful for separating languages)
    private String nameGen;

    // PATH PARAMETERS
    private final PathResolver pathResolver;

    @Setter
    private String resultMD5hash;

    // bounds of this map generated from polygon file
    @Getter
    private Boundaries boundary;

    @Getter
    @Setter
    private boolean isMerged = false;

    // MAIN PART

    public ItemMap(ItemMapPack mpParent) {
        super(mpParent);

         pathResolver = new PathResolver(this);

    }
//
//    public String getRelativeResultsPath() {
//        if (pathResult != null) {
//            int lastIndex = (Consts.DIR_BASE + "_result").length();
//            if (lastIndex == -1) {
//                return null;
//            }
//            return pathResult.toString().substring(lastIndex + 1);
//        }
//        return null;
//    }

    @Override
    public void validate() {
        // validate parent data
        super.validate();

        // check base parameters
        if (name == null || name.isEmpty()) {
            throw new IllegalArgumentException("Input XML is not valid. " +
                    "Invalid argument file: " + name);
        }

        // check country name
        if (hasAction(Action.ADDRESS_POI_DB) || hasAction(Action.GENERATE_MAPSFORGE)) {
            if (getCountryName() == null || getCountryName().isEmpty()) {
                throw new IllegalArgumentException("Input XML is not valid. " +
                        "Nor readable name nor country name is not defined for map or it's parent - name:" + name);
            }
        }
    }

    // GETTERS & SETTERS

    public Path getPathSource() {
        return pathResolver.getPath(PathType.EXTRACT, name + ".osm.pbf");
    }

    public Path getPathGenerate() {
        if (nameGen != null && !nameGen.isEmpty()) {
            return pathResolver.getPath(PathType.MAPSFORGE_GENERATE, nameGen + ".osm.map");
        }
        return pathResolver.getPath(PathType.MAPSFORGE_GENERATE, name + ".osm.map");
    }

    public Path getPathGenMlOutdoor() {
        return pathResolver.getPath(PathType.MAPLIBRE_ONLINE_OUTDOOR, name + "_lm_outdoor.mbtiles");
    }

    public Path getPathGenMlOpenMapTiles() {
        return pathResolver.getPath(PathType.MAPLIBRE_ONLINE_OPENMAPTILES, name + "_openmaptiles.mbtiles");
    }

    public Path getPathAddressPoiDb() {
        return pathResolver.getPath(PathType.ADDRESS_POI_DB, name + ".osm.db");
    }

    public Path getPathMerge() {
        return pathResolver.getPath(PathType.MERGE, name + ".osm.pbf");
    }

    public Path getPathPolygon() {
        return pathResolver.getPath(PathType.POLYGON, name + ".poly");
    }

    public Path getPathJsonPolygon() {
        return Utils.changeFileExtension(getPathPolygon(), ".json");
    }

    public Path getPathCountryBoundaryGeoJson() {
        return Utils.changeFileExtension(getPathPolygon(), "_country.geojson");
    }

    public Path getPathContour() {
        return pathResolver.getPath(PathType.CONTOUR, name + ".osm.pbf");
    }

    public Path getPathResult() {
        if (nameGen != null && !nameGen.isEmpty()) {
            return pathResolver.getPath(PathType.MAPSFORGE_RESULT, nameGen + ".zip");
        }
        return pathResolver.getPath(PathType.MAPSFORGE_RESULT, name + ".zip");
    }

    public Path getPathShp() {
        return pathResolver.getPath(PathType.SHP, name + ".shp");
    }

    public Path getPathCoastline() {
        return pathResolver.getPath(PathType.COASTLINE, name + ".osm.pbf");
    }

    public Path getPathTourist() {
        return pathResolver.getPath(PathType.TOURIST, name + ".osm.pbf");
    }

    public Path getPathTranform() {
        return pathResolver.getPath(PathType.TRANSFORM, name + ".osm.pbf");
    }

    // BASIC PARAMETERS

    /**
     * Get readable name of country in which is item. If country name is not defined returns readable name of map item
     *
     * @return name of country in readable form or empty string if is not defined
     */
    @Override
    public String getCountryName() {

        String countryName = super.getCountryName();
        if (countryName == null || countryName.isEmpty()) {
            return nameReadable;
        } else {
            return countryName;
        }
    }

    //  PARSE FUNCTIONS

    public void fillAttributes(KXmlParser parser) {
        // fill base parameters
        super.fillAttributes(parser);

        // set other private values
        if (parser.getAttributeValue(null, "id") != null) {
            id = parser.getAttributeValue(null, "id");
        }
        if (parser.getAttributeValue(null, "file") != null) {
            name = Utils.changeSlash(parser.getAttributeValue(null, "file"));
        }
        if (parser.getAttributeValue(null, "name") != null) {
            nameReadable = Utils.changeSlash(parser.getAttributeValue(null, "name"));
            if (nameReadable.isEmpty()) {
                Logger.w(TAG, "Config.xml not valid: Missing attribute name on line : " + parser.getLineNumber());
            }
        }
        if (parser.getAttributeValue(null, "fileGen") != null) {
            nameGen = Utils.changeSlash(parser.getAttributeValue(null, "fileGen"));
        }

        // test if MAP are valid
        validate();
    }

    // VARIOUS TOOLS

    public void setBoundsFromPolygon() throws IOException {

        File polyFile = getPathPolygon().toFile();
        if (!polyFile.exists()) {
            boundary = null;
            return;
        }

        double maxLatitude = -90.0;
        double maxLongitude = -180.0;
        double minLatitude = 90.0;
        double minLongitude = 180.0;

        //Scanner scan = null;
        BufferedReader br = null;
        try {
            //scan = new Scanner (new BufferedReader(new FileReader(polyFile)));
            br = new BufferedReader(new FileReader(polyFile));
            String line;
            while ((line = br.readLine()) != null) {
                //remove white space before and ond end of string line then
                // split string based on whitespace (regular expresion \\s+
                String[] cols = line.trim().split("\\s+");
                if (cols.length != 2) {
                    continue;
                }
                if (Utils.isNumeric(cols[0]) && Utils.isNumeric(cols[1])) {
                    double lon = Double.parseDouble(cols[0]);
                    double lat = Double.parseDouble(cols[1]);

                    maxLongitude = Math.max(lon, maxLongitude);
                    maxLatitude = Math.max(lat, maxLatitude);
                    minLongitude = Math.min(lon, minLongitude);
                    minLatitude = Math.min(lat, minLatitude);
                }
            }
            boundary = new Boundaries(minLongitude, maxLongitude, minLatitude, maxLatitude);
        } finally {
            if (br != null) {
                br.close();
            }
        }
    }

    /**
     * Read definition of map polygon from GeoJson
     */
    public JSONObject getItemAreaGeoJson() {

        // read json file with area definition
        File fileJsonPolyg = getPathJsonPolygon().toFile();
        if (!fileJsonPolyg.exists()) {
            throw new IllegalArgumentException("JSON polygon file doesn't exist " + fileJsonPolyg.getAbsolutePath());
        }

        String jsonPolygon = "";
        try {
            jsonPolygon = FileUtils.readFileToString(fileJsonPolyg, UTF_8);
        } catch (IOException e) {
            e.printStackTrace();
            throw new IllegalArgumentException("Can not read JSON polygon file " + fileJsonPolyg.getAbsolutePath());
        }
        // replace line brakes
        JSONParser parser = new JSONParser(net.minidev.json.parser.JSONParser.DEFAULT_PERMISSIVE_MODE);
        JSONObject obj = null;
        try {
            obj = (JSONObject) parser.parse(jsonPolygon);
        } catch (ParseException e) {
            e.printStackTrace();
        }

        return obj;
    }

    @Override
    public String toString() {
        return "ItemMap{" +
                "mId='" + id + '\'' +
                ", mName='" + name + '\'' +
                ", mNameGen='" + nameGen + '\'' +
                ", mResultMD5hash='" + resultMD5hash + '\'' +
                ", mBounds=" + boundary +
                ", isMerged=" + isMerged +
                '}';
    }
}

File: osmToolsBasic/src/main/java/com/asamm/osmTools/mapConfig/ItemMap.kt
package com.asamm.osmTools.mapConfig

import com.asamm.osmTools.config.Action
import com.asamm.osmTools.config.AppConfig
import com.asamm.osmTools.utils.Logger
import com.asamm.osmTools.utils.PolyUtils
import com.asamm.osmTools.utils.Utils
import net.minidev.json.JSONObject
import net.minidev.json.parser.JSONParser
import net.minidev.json.parser.ParseException
import org.apache.commons.io.FileUtils
import org.kxml2.io.KXmlParser
import java.io.BufferedReader
import java.io.File
import java.io.FileReader
import java.io.IOException
import java.nio.charset.StandardCharsets.UTF_8
import java.nio.file.Path

class ItemMap(parent: ItemMapPack?) : AItemMap(parent) {

    companion object {
        private val TAG = ItemMap::class.simpleName
    }

    /** Unique ID of this item */
    var id: String? = null
        private set

    /** Map file name (from the {@code file} XML attribute) */
    var fileName: String = ""
        private set

    /** Human-readable name shown in the store */
    var nameReadable: String = ""
        private set

    /** Alternative file name used during generating */
    private var nameGen: String = ""

    private val pathResolver = PathResolver(this)

    /** Bounds of this map, populated from the polygon file */
    var boundary: Boundaries? = null
        private set

    /** Returns true when this item represents the full planet extract */
    val isPlanet: Boolean
        get() = id != null && id == AppConfig.config.planetConfig.planetExtendedId

    /**
     * Country name with fallback: if not set on this item or its parent pack,
     * returns [nameReadable] so generators always have a usable country label.
     */
    override var countryName: String
        get() = super.countryName.ifEmpty { nameReadable }
        set(value) { super.countryName = value }

    // --- Validation ---

    override fun validate() {
        super.validate()
        require(fileName.isNotEmpty()) { "Input XML is not valid. Invalid argument file: $fileName" }
        if (hasAction(Action.ADDRESS_POI_DB) || hasAction(Action.GENERATE_MAPSFORGE)) {
            require(countryName.isNotEmpty()) {
                "Input XML is not valid. Country name not defined for map or its parent - name: $fileName"
            }
        }
    }

    // --- Parsing ---

    override fun fillAttributes(parser: KXmlParser) {
        super.fillAttributes(parser)
        parser.attr("id")?.let { id = it }
        parser.attr("file")?.let { fileName = Utils.changeSlash(it) }
        parser.attr("name")?.let {
            nameReadable = Utils.changeSlash(it)
            if (nameReadable.isEmpty()) {
                Logger.w(TAG, "Config.xml not valid: Missing attribute name on line: ${parser.lineNumber}")
            }
        }
        parser.attr("fileGen")?.let { nameGen = Utils.changeSlash(it) }
        validate()
    }

    // --- Paths ---

    /** Effective file name: uses [nameGen] when set, otherwise [fileName] */
    private val effectiveName: String get() = nameGen.ifEmpty { fileName }

    val pathSource: Path get() = pathResolver.getPath(PathType.EXTRACT, "$fileName.osm.pbf")
    val pathMapsforgeGenerate: Path get() = pathResolver.getPath(PathType.MAPSFORGE_GENERATE, "$effectiveName.osm.map")
    val pathMbtiles: Path get() = pathResolver.getPath(PathType.MBTILES_GENERATE, "$effectiveName.mbtiles")
    val pathAddressDb: Path get() = pathResolver.getPath(PathType.ADDRESS_DB, "$fileName.osm.db")

    /** Path to the legacy address POI database for LM Classic */
    val pathAddressPoiDb: Path get() = pathResolver.getPath(PathType.ADDRESS_POI_DB_CLASSIC, "$fileName.osm.db")

    val pathPoiV2Db: Path get() = pathResolver.getPath(PathType.POI_V2_DB, "$fileName.poiv2.db")
    val pathPolygon: Path get() = pathResolver.getPath(PathType.POLYGON, "$fileName.poly")
    val pathJsonPolygon: Path get() = Utils.changeFileExtension(pathPolygon, ".json")
    val pathCountryBoundaryGeoJson: Path get() = Utils.changeFileExtension(pathPolygon, "_country.geojson")
    val pathContour: Path get() = pathResolver.getPath(PathType.CONTOUR, "$fileName.osm.pbf")
    val pathResultMapsforge: Path get() = pathResolver.getPath(PathType.MAPSFORGE_RESULT, "$effectiveName.zip")
    val pathResultMapsforgeClassic: Path get() = Path.of(pathResultMapsforge.toString().replace(".zip", "_lmclassic.zip"))
    val pathResultMbtiles: Path get() = Path.of(pathResultMapsforge.toString().replace(".zip", "_mbtiles.zip"))
    val pathTourist: Path get() = pathResolver.getPath(PathType.TOURIST, "$fileName.osm.pbf")
    val pathResidential: Path get() = pathResolver.getPath(PathType.RESIDENTIAL, "$fileName.osm.pbf")
    val pathGenMlOutdoor: Path get() = pathResolver.getPath(
        PathType.MBTILES_ONLINE_OUTDOOR,
        "${if (isPlanet) "planet" else fileName}_lm_outdoor.mbtiles"
    )
    val pathGenPmtilesOnline: Path get() = pathResolver.getPath(
        PathType.PMTILES_ONLINE,
        "${if (isPlanet) "planet" else fileName}.pmtiles"
    )

    // --- Tools ---

    @Throws(IOException::class)
    fun setBoundsFromPolygon() {
        val polyFile = pathPolygon.toFile()
        if (!polyFile.exists()) {
            boundary = null
            return
        }

        var maxLat = -90.0;  var maxLon = -180.0
        var minLat = 90.0;   var minLon = 180.0

        BufferedReader(FileReader(polyFile)).use { br ->
            br.forEachLine { line ->
                val cols = line.trim().split("\\s+".toRegex())
                if (cols.size == 2 && Utils.isNumeric(cols[0]) && Utils.isNumeric(cols[1])) {
                    val lon = cols[0].toDouble()
                    val lat = cols[1].toDouble()
                    maxLon = maxOf(lon, maxLon); maxLat = maxOf(lat, maxLat)
                    minLon = minOf(lon, minLon); minLat = minOf(lat, minLat)
                }
            }
        }
        boundary = Boundaries(minLon, maxLon, minLat, maxLat)
    }

    /** Read the map polygon definition from GeoJSON, converting from .poly if needed */
    fun getItemAreaGeoJson(): JSONObject {
        val jsonFile = pathJsonPolygon.toFile()
        if (!jsonFile.exists()) {
            PolyUtils.polyFileToGeoJson(pathPolygon, jsonFile.toPath())
        }

        val jsonText = try {
            FileUtils.readFileToString(jsonFile, UTF_8)
        } catch (e: IOException) {
            Logger.e(TAG, "Can not read JSON polygon file ${jsonFile.absolutePath}", e)
            throw IllegalArgumentException("Can not read JSON polygon file ${jsonFile.absolutePath}")
        }

        return try {
            JSONParser(JSONParser.DEFAULT_PERMISSIVE_MODE).parse(jsonText) as JSONObject
        } catch (e: ParseException) {
            Logger.e(TAG, "Can not parse JSON polygon file ${jsonFile.absolutePath}", e)
            throw IllegalArgumentException("Can not parse JSON polygon file ${jsonFile.absolutePath}")
        }
    }

    override fun toString() =
        "ItemMap(id=$id, fileName=$fileName, nameGen=$nameGen, boundary=$boundary)"
}